### DISCUSSION TITLE
Is it possible to return an arbitrary Response from a form action?

### DISCUSSION THREAD

**Discussion by simonsarris** [sveltejs/kit#9948] [Q&A] [answered]
In an API GET I could write something to return an arbitrary Response, like:

```js
// API endpoint to get some file and initiate user download
export const GET = async ({ cookies }) => {
  // ...

  const someFile = fs.readFileSync(`someFilePathOnDrive`);
  return new Response(someFile, {
    status: 200,
    headers: {
      'Content-type': 'multipart/form-data',
      'Content-Disposition': `attachment; filename=someFileName.someFileExtension`
    }
  });
};
```

And it will download that file for the user. This file is not serializable.

Is it not possible in SvelteKit to do the same thing with form actions? Perhaps I'm simply overlooking something, or not understanding a limitation.

**Comment by teemingc** [comment]
Sounds like it could work if you throw a redirect to the `GET` endpoint at the end of the form action. (haven't tested this).
https://kit.svelte.dev/docs/load#redirects

Or would it work by simply using the `GET` endpoint URL as the form `action`? Instead of using the `POST` method and sveltekit's form actions.

**Comment by benquan** [comment]
@simonsarris Did you find a solution for your problem? I am trying to return a csv from a form action call.

**Comment by stephane-vanraes** [comment, marked as answer]
No it is not possible, actions are designed to populate the `form` property of a page and therefore have to be fully serializable.  A generic file, image or csv doesn't fall into that category. So the result from an action always has to be a simple object.

You will have to implement your form handling the "normal" way using an API endpoint as described here: https://kit.svelte.dev/docs/form-actions#alternatives

**Comment by simonsarris** [comment]
Yeah, I just ended up using an API endpoint.
